File: leetcode/sliding-window-maximum.cpp
/*
 * @Author: Tairan Gao
 * @Date:   2023-08-16 14:17:11
 * @Last Modified by:   Tairan Gao
 * @Last Modified time: 2023-08-16 14:29:38
 */

#include <vector>
#include <iostream>
#include <deque>

using namespace std;

template <typename T>
class MonotoniceQueue
{
private:
    size_t size;

public:
    deque<T> queue;
    MonotoniceQueue()
    {
        size = 0;
    }

    T get_max()
    {
        return queue.front();
    }

    void push_back(T element)
    {
        while (size > 0)
        {
            if (queue.back() < element)
            {
                queue.pop_back();
                size -= 1;
            }
            else
            {
                break;
            }
        }
        queue.push_back(element);
        size += 1;
    }

    void pop_max()
    {
        queue.pop_front();
        size -= 1;
    }
};

class Solution
{
public:
    vector<int> maxSlidingWindow(vector<int> &nums, int k)
    {

        MonotoniceQueue<int> queue;
        for (int i = 0; i < k; i++)
        {
            queue.push_back(nums[i]);
        }
        vector<int> sol({queue.get_max()});
        for (int i = k; i < nums.size(); i++)
        {
            int left = i - k;
            if (queue.get_max() == nums[left])
            {
                queue.pop_max();
            }
            queue.push_back(nums[i]);
            sol.push_back(queue.get_max());
        }
        return sol;
    }
};

int main()
{
    Solution s;
    vector<int> nums = {1};
    int k = 1;

    vector<int> res = s.maxSlidingWindow(nums, k);

    for (auto i : res)
    {
        cout << i << " ";
    }
    cout << endl;
}

File: leetcode/sliding-window-maximum.py
# -*- coding: utf-8 -*-
# @Author: Tairan Gao
# @Date:   2023-08-16 13:56:52
# @Last Modified by:   Tairan Gao
# @Last Modified time: 2023-08-16 14:16:12
from collections import deque
from typing import List


class MonotonicQueue:
    def __init__(self) -> None:
        self.queue = deque([])
        self.size = 0

    def push_back(self, element) -> None:
        while self.size > 0:
            if self.queue[-1] < element:
                self.queue.pop()
                self.size -= 1

            else:
                break
        self.queue.append(element)
        self.size += 1

    def get_max(self):
        return self.queue[0]

    def pop_max(self):
        self.queue.popleft()
        self.size -= 1


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        queue = MonotonicQueue()
        for i in range(k):
            queue.push_back(nums[i])

        sol = [queue.get_max()]

        for i in range(k, len(nums)):
            left = i - k
            if nums[left] == queue.get_max():
                queue.pop_max()
            queue.push_back(nums[i])
            sol.append(queue.get_max())

        return sol


if __name__ == "__main__":
    sol = Solution()
    nums = [1]
    k = 1
    print(sol.maxSlidingWindow(nums, k))
